src/blocktridiag.jl

```diff
@@ -0,0 +1,243 @@
+with sizes defind by `block_sizes`.
+
+```jldoctest
+julia> BlockTridiagMatrix(Matrix{Float64}, [1,3], [2,2])
+2×2-blocked 4×4 BlockArrays.BlockTridiagMatrix{Float64,2,Array{Float64,2}}:
+ #undef  │  #undef  #undef  #undef  │
+ --------┼--------------------------┼
+ #undef  │  #undef  #undef  #undef  │
+ #undef  │  #undef  #undef  #undef  │
+ --------┼--------------------------┼
+ #undef  │  #undef  #undef  #undef  │
+```
+"""
+@inline function BlockTridiagMatrix{T, R <: AbstractMatrix{T}
+                                   }(::Type{R},
+                                     block_sizes::Vararg{Vector{Int}, 2})
+    BlockTridiagMatrix(R, BlockSizes(block_sizes...))
+end
+
+function BlockTridiagMatrix{T, R <: AbstractMatrix{T}}(::Type{R}, block_sizes::BlockSizes{2})
+    n_blocks = nblocks(block_sizes)
+    n_blocks[1] == n_blocks[2] || throw("expect same number of blocks in both dimensions")
+    diagl = Vector{R}(n_blocks[1])
+    lower = Vector{R}(n_blocks[1]-1)
+    upper = Vector{R}(n_blocks[1]-1)
+    BlockTridiagMatrix{T,R}(diagl, lower, upper, block_sizes)
+end
+
+function BlockTridiagMatrix(arr::AbstractMatrix,
+                            block_sizes::Vararg{Vector{Int}, 2})
+    for i in 1:2
+        if sum(block_sizes[i]) != size(arr, i)
+            throw(DimensionMismatch(
+                    "block size for dimension $i: $(block_sizes[i])" *
+                    "does not sum to the array size: $(size(arr, i))"))
+        end
+    end
+
+    _block_sizes = BlockSizes(block_sizes...)
+    bltrid_mat = BlockTridiagMatrix(typeof(arr), _block_sizes)
+    row_blocks, col_blocks = nblocks(bltrid_mat)
+    for brow in 1:row_blocks
+        for bcol in max(1,brow-1):min(brow+1,col_blocks)
+            indices = globalrange(_block_sizes, (brow,bcol))
+            setblock!(bltrid_mat, arr[indices...], brow, bcol)
+        end
+    end
+
+    return bltrid_mat
+end
+
+################################
+# AbstractBlockArray Interface #
+################################
+@inline nblocks(bltrid_mat::BlockTridiagMatrix) = nblocks(bltrid_mat.block_sizes)
+@inline blocksize(bltrid_mat::BlockTridiagMatrix, i::Int, j::Int) = blocksize(bltrid_mat.block_sizes, (i, j))
+
+@inline function getblock(bltrid_mat::BlockTridiagMatrix, i::Int, j::Int)
+    @boundscheck blockcheckbounds(bltrid_mat, i, j)
+    if i==j
+        # for blocks on the diagonal,
+        # get the block from `diagl`
+        return bltrid_mat.diagl[i]
+    elseif i==j+1
+        # for blocks below the diagonal,
+        # get the block from `lower`
+        return bltrid_mat.lower[j]
+    elseif i+1==j
+        # for blocks above the diagonal,
+        # get the block from `upper`
+        return bltrid_mat.upper[i]
+    else
+        # otherwise return a freshly-baked
+        # matrix of zeros (with a warning
+        # because that's dumb)
+        warn(@sprintf("""The (%d,%d) block of a block tridiagonal matrix
```

Comment:
I don't like warnings like this. It should either be an error or be allowed without warning. Potential wastefulness should go in the documentation of the type. You don't get a warning if you access a zero element of a sparse matrix.